src/hedera/hederaEngine.js

```diff
@@ -0,0 +1,404 @@
+/**
+ * Created by Austin Bonander <austin@launchbadge.com> on 9/30/19.
+ */
+// @flow
+
+import * as hedera from '@hashgraph/sdk'
+import baseX from 'base-x'
+import { bns } from 'biggystring'
+import { EdgeFreshAddress } from 'edge-core-js'
+import {
+  type EdgeCurrencyEngineOptions,
+  type EdgeCurrencyTools,
+  type EdgeIo,
+  type EdgeSpendInfo,
+  type EdgeTransaction,
+  type EdgeWalletInfo,
+  NoAmountSpecifiedError
+} from 'edge-core-js/types'
+
+import { CurrencyEngine } from '../common/engine.js'
+import { hexToBuf } from '../common/utils'
+import { useTestnet } from './hederaInfo'
+import { HederaPlugin } from './hederaPlugin'
+
+const base64Digits =
+  'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz1234567890+/'
+
+const base64 = baseX(base64Digits)
+
+export class HederaEngine extends CurrencyEngine {
+  hederaPlugin: HederaPlugin
+  client: hedera.Client
+  accountId: ?hedera.AccountId
+  otherMethods: Object
+  io: EdgeIo
+  creatorApiServer: string
+
+  constructor(
+    currencyPlugin: HederaPlugin,
+    walletInfo: EdgeWalletInfo,
+    opts: EdgeCurrencyEngineOptions,
+    io: EdgeIo
+  ) {
+    super(currencyPlugin, walletInfo, opts)
+    this.log('hedera constructor called', walletInfo)
+    this.hederaPlugin = currencyPlugin
+
+    this.io = io
+    this.client = useTestnet
+      ? hedera.Client.forTestnet()
+      : hedera.Client.forMainnet()
+
+    this.creatorApiServer = this.currencyInfo.defaultSettings.otherSettings.creatorApiServers[0]
+
+    this.otherMethods = {
+      getAccountActivationQuote: async (params: Object) => {
+        const { currencyCode, activePublicKey } = params
+        if (!currencyCode || !activePublicKey) {
+          throw new Error('ErrorInvalidParams')
+        }
+
+        const options = {
+          method: 'POST',
+          headers: {
+            Accept: 'application/json',
+            'Content-Type': 'application/json'
+          },
+          body: JSON.stringify({
+            currency: currencyCode,
+            public_key: activePublicKey
+          })
+        }
+
+        const response = await (
+          await io.fetch(`${this.creatorApiServer}/account`, options)
+        ).json()
+
+        if (response.status === 'success') {
+          this.walletLocalData.otherData.activationPaymentUrl =
+            response.payment_url
+          this.walletLocalDataDirty = true
+
+          return {
+            paymentAddress: response.address,
+            currencyCode,
+            amount: bns.toFixed(response.amount, 3, 9),
+            exchangeAmount: '0'
+          }
+        }
+
+        throw new Error({
+          message: 'ErrorActivationRequest',
+          detail: response.message
+        })
```

Comment:
Do you have a convention for passing additional context with an error? I could have sworn I saw this pattern elsewhere in the codebase but I could be misremembering.

Reply:
@swansontec Can you comment on the best method of passing errors. I know you thought about this one a lot.